EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+
+1. Jumpdest analysis is only run on the *code*.
+2. Execution starts at the first byte of the *code*, and `PC` is set to this position within the container format (e.g. `PC=10` for a *container* with a code and data section).
+3. If `PC` goes outside of the code section bounds, execution  aborts with failure.
+4. `PC` returns the current position within the *container*.
+5. `JUMP`/`JUMPI` uses an absolute offset within the *container*.
+6. `CODECOPY`/`CODESIZE`/`EXTCODECOPY`/`EXTCODESIZE`/`EXTCODEHASH` keeps operating on the entire *container*.
+7. The input to `CREATE`/`CREATE2` is still the entire *container*.
+
+#### Changes to contract creation semantics
+
+For clarity, the *EOF prefix* together with a version number $v$ is denoted as the *EOF$v$ prefix*, e.g. *EOF1 prefix*. 
+
+1. If _initcode's container_ has EOF1 prefix it must be valid EOF1 code.
+2. If _code's container_ has EOF1 prefix it must be valid EOF1 code.
+
+## Rationale
+
+EVM and/or account versioning has been discussed numerous times over the past years. This proposal aims to learn from them. A good starting point collecting the previous proposals is [this thread](https://ethereum-magicians.org/t/ethereum-account-versioning/3508).
+
+### Execution vs. creation time validation
+
+This specification introduces creation time validation, which means:
+- All created contracts with EOF$n$ prefix are valid according to version $n$ rules. This is very strong and useful property. The client can trust that the deployed code is well-formed.
+- In future, this allows serializing `JUMPDEST` map in the EOF container and eliminate the need of implicit `JUMPDEST` analysis required before execution.
+- Or completely remove the need for `JUMPDEST` instructions.
+- This helps with deprecating EVM instructions and/or features.
+- The biggest disadvantage is that deploy-time validation of EOF code must be enabled in two hard-forks.
+
+The alternative is to have execution time validation for EOF. This is performed every single time a contract is executed, however clients may be able to cache validation results. This approach has the following properties:
+- Because the validation is consensus-level execution step, it means the execution always requires the entire code. This makes code merkleization impractical.
+- The main advantage is it can be enabled via a single hard-fork.
+- A second advantage is better backwards compatibility: data contracts starting with the `0xEF` byte or the *EOF prefix* can be deployed. This is a dubious benefit however.
+
+### Initcode vs code
+
+After HF1 and before HF2 the `FORMAT` first byte check only applies to _code_. Applying the rule also to _initcode_ would have the same effects because is not distinguishable from executing _initcode_: if it starts with `FORMAT` execution would exceptionally abort. Yet, we decided against introducing an explicit check for _initcode_ for the simplicity of the HF1 spec.
```

Comment:
Again, this EIP should be written standalone and without regard to the hardfork coordination process.